[SVG2] Add support for the 'turn' unit in <angle>

karlcow opened the pull request:
#### 82c987e62f88df8ecd31f4526b688c0bbfdba45d
<pre>
[SVG2] Add support for the &apos;turn&apos; unit in &lt;angle&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=267180">https://bugs.webkit.org/show_bug.cgi?id=267180</a>
<a href="https://rdar.apple.com/120840743">rdar://120840743</a>

Reviewed by Anne van Kesteren.

This is adding the turn units to SVGAngle.
Note that the spec is saying that the IDL should not be modified.
<a href="https://svgwg.org/svg2-draft/types.html#__svg__SVGAngle__SVG_ANGLETYPE_UNKNOWN">https://svgwg.org/svg2-draft/types.html#__svg__SVGAngle__SVG_ANGLETYPE_UNKNOWN</a>
And new units should be considered with the SVG_ANGLETYPE_UNKNOWN.

* Source/WTF/wtf/MathExtras.h:
(turn2grad):
(grad2turn):
* Source/WebCore/svg/SVGAngle.h:
(WebCore::SVGAngle::unitType const):
* Source/WebCore/svg/SVGAngleValue.cpp:
(WebCore::SVGAngleValue::value const):
(WebCore::SVGAngleValue::setValue):
(WebCore::SVGAngleValue::valueAsString const):
(WebCore::parseAngleType):
(WebCore::SVGAngleValue::newValueSpecifiedUnits):
(WebCore::SVGAngleValue::convertToSpecifiedUnits):
* Source/WebCore/svg/SVGAngleValue.h:

Co-authored-by: Anne van Kesteren &lt;annevk@annevk.nl&gt;
Canonical link: <a href="https://commits.webkit.org/274778@main">https://commits.webkit.org/274778@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/130e0201c49ca213c7f7aae2ecbc3b7ee998b191

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39901 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18912 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42276 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42445 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35803 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42208 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21819 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16242 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33259 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40475 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15946 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34502 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13803 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/13827 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35458 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43724 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33354 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36235 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35794 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39560 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39527 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14716 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12110 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37831 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16335 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46535 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8974 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16384 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9573 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15979 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/svg/SVGAngleValue.h:
```diff
@@ -34,7 +34,8 @@ class SVGAngleValue {
         SVG_ANGLETYPE_UNSPECIFIED = 1,
         SVG_ANGLETYPE_DEG = 2,
         SVG_ANGLETYPE_RAD = 3,
-        SVG_ANGLETYPE_GRAD = 4
+        SVG_ANGLETYPE_GRAD = 4,
+        SVG_ANGLETYPE_TURN = 5
```
According to the style error, it seems I need to change the casing style of the enum type. 

```
UnspecifiedSvgAngleType = 1,
DegSvgAngleType = 2,
RadSvgAngleType = 3,
GradSvgAngleType = 4,
TurnSvgAngleType = 5
```

something like this ?

nt1m replied:
If I were to clean up this code separately, I would switch to:
```
enum class Type : uint8_t {
  Unknown,
  Deg,
  Rad,
  Grad,
  Turn,
}
```

Definitely a separate task though.

karlcow replied:
I added a FIXME in the source code.

Ahmad-S792 replied:
@karlcow - Should we add this bug link - https://bugs.webkit.org/show_bug.cgi?id=269429

karlcow replied:
@nt1m How would you solve the difference with IDL?

https://searchfox.org/wubkat/source/Source/WebCore/svg/SVGAngleValue.h
https://searchfox.org/wubkat/source/Source/WebCore/svg/SVGAngle.idl

@Ahmad-S792 Yes!

karlcow replied:
@Ahmad-S792 That said I added a pointer to the discussion here.

Ahmad-S792 replied:
Should we raise this issue to update IDL in SVGWG GitHub?

karlcow replied:
This would be harder to change, given that this is implemented and available on the Web.

Ahmad-S792 commented on Source/WebCore/svg/SVGAngleValue.cpp:
```diff
@@ -134,13 +143,31 @@ ExceptionOr<void> SVGAngleValue::newValueSpecifiedUnits(unsigned short unitType,
 ExceptionOr<void> SVGAngleValue::convertToSpecifiedUnits(unsigned short unitType)
 {
-    if (unitType == SVG_ANGLETYPE_UNKNOWN || m_unitType == SVG_ANGLETYPE_UNKNOWN || unitType > SVG_ANGLETYPE_GRAD)
+    if (unitType == SVG_ANGLETYPE_UNKNOWN || m_unitType == SVG_ANGLETYPE_UNKNOWN || unitType > SVG_ANGLETYPE_GRAD || unitType > SVG_ANGLETYPE_TURN)
```
I missed this in my local patch. Good spot!

Ahmad-S792 commented on Source/WebCore/svg/SVGAngle.h:
```diff
@@ -56,6 +56,11 @@ class SVGAngle : public SVGValueProperty<SVGAngleValue> {
+        // > If other types of angles are supported and used, the SVGAngle uses
+        // > the SVG_ANGLETYPE_UNKNOWN unit type.
+        if (m_value.unitType() > 4)
+            return 0;
```
Might be to return ```return SVGAngleValue::SVG_ANGLETYPE_UNKNOWN;```?

karlcow replied:
yes. We just fixed it, with @annevk

annevk commented on Source/WebCore/svg/SVGAngleValue.cpp:
```diff
@@ -125,7 +134,7 @@ ExceptionOr<void> SVGAngleValue::setValueAsString(const String& value)
 ExceptionOr<void> SVGAngleValue::newValueSpecifiedUnits(unsigned short unitType, float valueInSpecifiedUnits)
 {
-    if (unitType == SVG_ANGLETYPE_UNKNOWN || unitType > SVG_ANGLETYPE_GRAD)
+    if (unitType == SVG_ANGLETYPE_UNKNOWN || unitType > SVG_ANGLETYPE_TURN)
```
I don't think you should have changed this or below. These methods can't know about `SVG_ANGLETYPE_TURN` after all. (These are exposed through IDL.)

karlcow replied:
indeed… damned.
https://svgwg.org/svg2-draft/types.html#InterfaceSVGAngle

Step 2 in https://svgwg.org/svg2-draft/types.html#__svg__SVGAngle__SVG_ANGLETYPE_UNKNOWN

> Otherwise, return [SVG_ANGLETYPE_UNKNOWN](https://svgwg.org/svg2-draft/types.html#__svg__SVGAngle__SVG_ANGLETYPE_UNKNOWN).
> For example, for an [`<angle>`](https://www.w3.org/TR/css-values/#angles) with a turn unit, [SVG_ANGLETYPE_UNKNOWN](https://svgwg.org/svg2-draft/types.html#__svg__SVGAngle__SVG_ANGLETYPE_UNKNOWN) would be returned.

annevk commented on Source/WebCore/svg/SVGAngleValue.cpp:
```diff
@@ -135,13 +144,31 @@ ExceptionOr<void> SVGAngleValue::newValueSpecifiedUnits(unsigned short unitType,
 ExceptionOr<void> SVGAngleValue::convertToSpecifiedUnits(unsigned short unitType)
 {
-    if (unitType == SVG_ANGLETYPE_UNKNOWN || m_unitType == SVG_ANGLETYPE_UNKNOWN || unitType > SVG_ANGLETYPE_GRAD)
+    if (unitType == SVG_ANGLETYPE_UNKNOWN || m_unitType == SVG_ANGLETYPE_UNKNOWN || unitType > SVG_ANGLETYPE_TURN)
```
Same here.

annevk commented on Source/WebCore/svg/SVGAngleValue.cpp:
```diff
@@ -96,6 +103,8 @@ template<typename CharacterType> static inline SVGAngleValue::Type parseAngleTyp
     case 4:
         if (compareCharacters(buffer.position(), 'g', 'r', 'a', 'd'))
             return SVGAngleValue::SVG_ANGLETYPE_GRAD;
+        if (buffer[0] == 't' && buffer[1] == 'u' && buffer[2] == 'r' && buffer[3] == 'n')
```
This should use compareCharacters().

karlcow replied:
fixed

annevk commented on Source/WebCore/svg/SVGAngleValue.h:
```diff
@@ -30,11 +30,14 @@ class SVGAngleValue {
     WTF_MAKE_FAST_ALLOCATED;
 public:
     enum Type {
+        // FIXME: change the casing style of the enum type or naming of the Angle Value
```
FIXMEs have a particular style, they need to be written as:

"FIXME: Sentence starts with uppercase and ends with a dot."

And then when there's a URL they're written as

"FIXME: We need to fix this (url)."

karlcow replied:
fixed.

annevk commented on Source/WebCore/svg/SVGAngle.h:
```diff
@@ -56,6 +56,11 @@ class SVGAngle : public SVGValueProperty<SVGAngleValue> {
     {
+        // Per spec https://svgwg.org/svg2-draft/types.html#__svg__SVGAngle__SVG_ANGLETYPE_UNKNOWN
+        // > If other types of angles are supported and used, the SVGAngle uses
+        // > the SVG_ANGLETYPE_UNKNOWN unit type.
```
Typically we don't include long comments in the code. A URL is okay though.

karlcow replied:
fixed

annevk reviewed (approved):
r+ modulo nits.

karlcow committed:
[SVG2] Add support for the 'turn' unit in <angle>
https://bugs.webkit.org/show_bug.cgi?id=267180
rdar://120840743

Reviewed by Anne van Kesteren.

This is adding the turn units to SVGAngle.
Note that the spec is saying that the IDL should not be modified.
https://svgwg.org/svg2-draft/types.html#__svg__SVGAngle__SVG_ANGLETYPE_UNKNOWN
And new units should be considered with the SVG_ANGLETYPE_UNKNOWN.

* Source/WTF/wtf/MathExtras.h:
(turn2grad):
(grad2turn):
* Source/WebCore/svg/SVGAngle.h:
(WebCore::SVGAngle::unitType const):
* Source/WebCore/svg/SVGAngleValue.cpp:
(WebCore::SVGAngleValue::value const):
(WebCore::SVGAngleValue::setValue):
(WebCore::SVGAngleValue::valueAsString const):
(WebCore::parseAngleType):
(WebCore::SVGAngleValue::newValueSpecifiedUnits):
(WebCore::SVGAngleValue::convertToSpecifiedUnits):
* Source/WebCore/svg/SVGAngleValue.h:

Co-authored-by: Anne van Kesteren <annevk@annevk.nl>
Canonical link: https://commits.webkit.org/274778@main
